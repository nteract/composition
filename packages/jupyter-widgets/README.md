# `transform-widgets`

Rich media type which will render Jupyter-Widgets in isolated iframes.

`MIMETYPE = "application/vnd.jupyter.widget-view+json"`

## Usage

### Within nteract

```jsx
import { WidgetDisplay } from "@nteract/transform-widgets";

...

<WidgetDisplay data={{model_id: 'widget-model-guid'}}/>
```

### Outside of nteract

```jsx
import { PureWidgetDisplay } from "@nteract/transform-widgets";

...

<PureWidgetDisplay 
    data={{model_id: 'widget-model-guid'} 
    currentKernel={remoteKernelPropsInstance}}/>
```

See `RemoteKernelProps` of `nteract/packages/core/src/state/entities/kernels.js` 
for a description of `remoteKernelPropsInstance`.
