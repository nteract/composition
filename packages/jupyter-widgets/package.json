{
  "name": "@nteract/jupyter-widgets",
  "version": "1.0.0-alpha.0",
  "description": "Transform for rendering Jupyter-JS-Widgets in output areas.",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "nteractDesktop": "src/index.ts",
  "scripts": {
    "prepare": "npm run build",
    "prepublishOnly": "npm run:build",
    "build": "npm run build:clean && npm run build:lib",
    "build:clean": "tsc -b --clean",
    "build:lib": "tsc -b",
    "build:lib:watch": "tsc -b --watch",
    "build:watch": "npm run build:clean && npm run build:lib:watch",
    "test": "jest ."
  },
  "repository": "https://github.com/nteract/nteract/tree/master/packages/transforms",
  "keywords": [
    "nteract",
    "transforms",
    "notebook"
  ],
  "author": "Jonathan Frederic <jfrederic@google.com>",
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/nteract/nteract/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@nteract/messaging": "^5.0.0-alpha.0",
    "ansi-to-react": "^4.0.0-alpha.0"
  },
  "peerDependencies": {
    "react": "^16.3.2"
  }
}
