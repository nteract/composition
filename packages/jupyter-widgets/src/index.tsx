import * as React from "react";
import {Subject} from "rxjs";

// import { OuterShim } from "./outer-shim";

type Props = {
  data: { model_id: string },
  channels: Subject<any>
};

/**
 * Component used to render a widget view.
 *
 * Given a kernel and widget model id, this component will render a widget
 * view for the widget model. The widget view and widget models are isolated to
 * a cross domain iframe. Since they have access to the kernel, they are capable
 * of communicating directly with the kernel instance which makes them atypical.
 *
 * Even though it may appear to be pure, since it doesn't have react state, this
 * component's iframe maintains it's own state in communication with the kernel.
 */
export class WidgetDisplay extends React.Component<Props, null> {
  static MIMETYPE = "application/vnd.jupyter.widget-view+json";

  // TODO: Uncomment this and related code in a follow-up PR.
  // The outer shim is responsable for managing the rendered cross domain
  // iframe, communicating with it, and relaying information to and from the
  // kernel.
  // private shim: OuterShim;

  // A reference to the div which we can inject the cross domain widget iframe.
  private container = React.createRef<HTMLDivElement>();

  /** @override */
  componentDidMount() {
    if (!this.container.current) return;

    // When the component has mounted, inject the widget iframe.
    // const view = this.shim.getView();
    // this.container.current.appendChild(view);
  }

  /** @override */
  componentWillUnmount() {
    // if (this.shim) {
    //   this.shim.dispose();
    //   delete this.shim;
    // }
  }

  /** @override */
  shouldComponentUpdate(nextProps: Props): boolean {
    // Only update if the model_id or kernel_id have changed.
    if (nextProps.data.model_id !== this.props.data.model_id) {
      return true;
    }

    // If the channels have changed, we need to make sure that we connect the
    // shim to the new channels.
    if (nextProps.channels !== this.props.channels) {
      return true;
    }

    return false;
  }

  /** @override */
  render() {
    this.createOrUpdateShim();

    return (
      <pre>
        Jupyter-Widgets are not yet supported in nteract.
        <div ref={this.container} />
      </pre>
    );
  }

  /**
   * Creates or updates the existing shim to the current model id and kernel.
   */
  private createOrUpdateShim() {
    // if (!this.shim) {
    //   this.shim = new OuterShim();
    // }
    // this.shim.setCommMsgsSubject(this.props.channels);
    // this.shim.setModelId(this.props.data.model_id);
  }
}
