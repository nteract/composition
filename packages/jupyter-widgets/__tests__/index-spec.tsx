import * as React from "react";
import { WidgetDisplay } from "@nteract/jupyter-widgets";
import { Subject } from "rxjs";
import * as renderer from "react-test-renderer";

describe('index', () => {
  describe('WidgetDisplay', () => {
    /**
     * Tests that the component can be constructed.
     */
    it('can be constructed', () => {
      expect(<WidgetDisplay
        data={{model_id: 'none'}}
        channels={new Subject()}/>).not.toBeNull();
    });

    /**
     * Tests that the component can be rendered.
     */
    it('matches snapshot', () => {
      const renderedComponent = renderer.create(
          <WidgetDisplay data={{model_id: 'none'}} channels={new Subject()}/>);
      expect(renderedComponent.toJSON()).toMatchSnapshot();
    });
  });
});