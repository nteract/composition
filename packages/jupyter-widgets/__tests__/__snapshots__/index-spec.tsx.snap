// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index WidgetDisplay matches snapshot 1`] = `
<pre>
  Jupyter-Widgets are not yet supported in nteract.
  <div />
</pre>
`;
